src/Symfony/Component/Console/Application.php

```diff
@@ -992,19 +1005,34 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
             if ($this->dispatcher) {
                 // We register application signals, so that we can dispatch the event
                 foreach ($this->signalsToDispatchEvent as $signal) {
-                    $event = new ConsoleSignalEvent($command, $input, $output, $signal);
-
-                    $this->signalRegistry->register($signal, function ($signal) use ($event, $command, $commandSignals) {
-                        $this->dispatcher->dispatch($event, ConsoleEvents::SIGNAL);
-                        $exitCode = $event->getExitCode();
+                    $signalEvent = new ConsoleSignalEvent($command, $input, $output, $signal);
+                    $alarmEvent = \SIGALRM === $signal ? new ConsoleAlarmEvent($command, $input, $output) : null;
+
+                    $this->signalRegistry->register($signal, function ($signal) use ($signalEvent, $alarmEvent, $command, $commandSignals, $input, $output) {
+                        $this->dispatcher->dispatch($signalEvent, ConsoleEvents::SIGNAL);
+                        $exitCode = $signalEvent->getExitCode();
+
+                        if (null !== $alarmEvent) {
+                            if (false !== $exitCode) {
+                                $alarmEvent->setExitCode($exitCode);
+                            } else {
+                                $alarmEvent->abortExit();
+                            }
+                            $this->dispatcher->dispatch($alarmEvent, ConsoleEvents::ALARM);
+                            $exitCode = $alarmEvent->getExitCode();
+                        }
 
+                        if (\SIGALRM === $signal && $command instanceof AlarmableCommandInterface) {
+                            $exitCode = $command->handleAlarm($exitCode);
+                            $this->signalRegistry->scheduleAlarm($command->getAlarmTime($input, $output));
+                        }
                         // If the command is signalable, we call the handleSignal() method
-                        if (\in_array($signal, $commandSignals, true)) {
+                        elseif (\in_array($signal, $commandSignals, true)) {
```

Comment:
this will never be called in dev when a SignalableCommand asks for `SIGALRM` because of the TraceableCommand wrapping it and implementing AlarmableCommandInterface. This means this PR is currently a BC break.